### PR TITLE
 [ENH] interface the seasonal package as a parameter estimator for seasonality

### LINKED ISSUE #3835: [ENH] interface the `seasonal` package as a parameter estimator for seasonality (`BaseParamFitter` descendant)
We now have an interface design for generic parameter estimators, with `BaseParamFitter`.
(original design issue: https://github.com/sktime/sktime/issues/1056)

Still experimental, we should further explore it by interfacing useful parameter estimators.

@mloning suggested the methods in the `seasonal` package in https://github.com/sktime/sktime/issues/1056, given the existing base class interface and seasonality examples it seems like a good first issue.

### PR DESCRIPTION
**Pull request by blazingbhavneek**
Fixes #3835

Interfaces welch/seasonal package as seasonality param_est

Introduces soft dependancey: [Seasonal](https://github.com/welch/seasonal)

### CODE REVIEW

**Comment by blazingbhavneek** [comment]
@achieveordie Can you tell me what is the issue here? (Last time you told me to change pyproject.toml)

File: sktime/param_est/seasonality.py [hunk 1]
```diff
@@ -2,10 +2,11 @@
 # copyright: sktime developers, BSD-3-Clause License (see LICENSE file)
 """Parameter estimators for seasonality."""
 
-__author__ = ["fkiraly"]
-__all__ = ["SeasonalityACF"]
+__author__ = ["fkiraly", "blazingbhavneek"]
+__all__ = ["SeasonalityACF", "Seasonality"]
 
 import numpy as np
+from seasonal.periodogram import periodogram
```
Comment (achieveordie) [inline sktime/param_est/seasonality.py:9]: Soft imports should not be in the scope of the file but should rather be limited to the method(s) that use it. Take a look at tensorflow/keras in the DL submodule for an example of how to do that (It involves simply importing it within the method where this package is used.)

The reason behind this is simple - being a soft dependency, one might not have installed it but may still want to use a function/object (and this function/object should not be making use of this dependency). A good software design will not throw an error for an import which isn't supposed to be used in the code execution.

File: sktime/param_est/seasonality.py [hunk 2]
```diff
@@ -407,3 +406,93 @@ def get_test_params(cls, parameter_set="default"):
...
+
+        Returns
+        -------
+        params : dict or list of dict, default = {}
+            Parameters to create testing instances of the class
+            Each dict are parameters to construct an "interesting" test instance, i.e.,
+            `MyClass(**params)` or `MyClass(**params[i])` creates a valid test instance.
+            `create_test_instance` uses the first (or only) dictionary in `params`
+        """
+        params1 = {}
```
Comment (achieveordie) [inline sktime/param_est/seasonality.py:495]: Why leave it empty? Unless I'm misunderstanding something, the entire purpose of this method is to overwrite default params to facilitate faster testing. I suggest making use of a set of params that maximizes code coverage.

**Review by achieveordie — CHANGES_REQUESTED** [review]

File: sktime/param_est/seasonality.py [hunk 3]
```diff
@@ -141,7 +141,7 @@ def _fit(self, X):
 
         candidate_sp = self.candidate_sp
         if candidate_sp is None:
-            candidate_sp = range(2, nlags + 1)
+            candidate_sp = range(2, nlags)
```
Comment (ngupta23) [inline sktime/param_est/seasonality.py:144]: Why is this being reverted back to nlags? nlags + 1 was added for a specific reason?

@fkiraly FYI...

Comment (fkiraly) [reply sktime/param_est/seasonality.py:144]: uh-oh, I think this is just an accident, probably from file copy-paste.
Should be reverted.

File: sktime/param_est/seasonality.py (hunk 1 above)
Comment (blazingbhavneek) [reply sktime/param_est/seasonality.py]: Done!

File: sktime/param_est/seasonality.py (hunk 2 above)
Comment (blazingbhavneek) [reply sktime/param_est/seasonality.py]: I see I made a few mistakes here

1. There are no parameters required for this class, so we don't need them anyways
2. I have not written any tests for it either (simple interfacing task)

(Copied Franz Kiraly's code for reference, but forgot to delete this part)

Comment (fkiraly) [reply sktime/param_est/seasonality.py]: it has `min_period` and `max_period`, I would expose these as parameters, possibly through `candidate_sp`

File: sktime/param_est/seasonality.py [hunk 4]
```diff
@@ -407,3 +405,71 @@ def get_test_params(cls, parameter_set="default"):
         params2 = {"candidate_sp": [3, 7, 12]}
 
         return [params1, params2]
+
+
+class Seasonality(BaseParamFitter):
```
Comment (fkiraly) [inline sktime/param_est/seasonality.py:410]: I wouldn´t call it `Seasonality`, maybe be more specific, e.g., `SeasonalityPeriodogram` or similar

Comment (blazingbhavneek) [reply sktime/param_est/seasonality.py]: Done!

File: sktime/param_est/seasonality.py (hunk 2 above)
Comment (blazingbhavneek) [reply sktime/param_est/seasonality.py:530]: Done! I also wrote test parameters for it, will write tests for it later

File: sktime/param_est/seasonality.py (hunk 4 above)
Comment (fkiraly) [reply sktime/param_est/seasonality.py]: thanks

**Review by fkiraly — CHANGES_REQUESTED** [review]
Thanks!

Can you please revert the merge accident that @ngupta23 pointed out?

I.e., please do not make any changes to the existing estimator, `SeasonalityACF`.

**Comment by fkiraly** [comment]
I see the commit, but that weren't all changes. I would suggest, just copy `SeasonalityACF` from `main`.

File: sktime/param_est/seasonality.py [hunk 5]
```diff
@@ -407,3 +407,130 @@ def get_test_params(cls, parameter_set="default"):
...
+
+
+class SeasonalityPeriodogram(BaseParamFitter):
+    """Score periodicities by their spectral power.
+
+    Interfacing `seasonal.periodogram` to determine candidate seasonality parameters.
+
+    Parameters
+    ----------
+    candidate_sp : None, int or list of int, optional, default = None
```
Comment (fkiraly) [inline sktime/param_est/seasonality.py:419]: from the code, it looks like the default is 4?

Further, I think the mapping of `candidate_sp` on `min_period` and `max_period` might be confusing to the user.
Perhaps you just want to patch through `min_period` and `max_period`, that might be clearer?

File: sktime/param_est/seasonality.py [hunk 6]
```diff
@@ -407,3 +407,110 @@ def get_test_params(cls, parameter_set="default"):
...
+        Defaults to 4
+    max_period : int
+        Disregard periods longer than this number of samples.
+        Defaults to None
+
+    Attributes
+    ----------
+    sp_ : int, seasonality period at lowest p-level, if any sub-threshold, else 1
+        if `candidate_sp` is passed, will be in `candidate_sp` or 1
+    sp_significant_ : list of int, seasonality periods with sub-threshold p-levels
```
Comment (fkiraly) [inline sktime/param_est/seasonality.py:430]: is this correct, i.e., are they ordered by p-level? If not, the docstring needs to be changed.

File: sktime/param_est/seasonality.py [hunk 7]
```diff
@@ -407,3 +407,110 @@ def get_test_params(cls, parameter_set="default"):
...
+        Disregard periods shorter than this number of samples.
+        Defaults to 4
+    max_period : int
+        Disregard periods longer than this number of samples.
+        Defaults to None
+
+    Attributes
+    ----------
+    sp_ : int, seasonality period at lowest p-level, if any sub-threshold, else 1
+        if `candidate_sp` is passed, will be in `candidate_sp` or 1
```
Comment (fkiraly) [inline sktime/param_est/seasonality.py:429]: there is no argument `candidate_sp`, docstring needs to be adapted

**Comment by blazingbhavneek** [comment]
@fkiraly @achieveordie Requested changes are addressed as far as I know. Am I missing something?

**Review by achieveordie — APPROVED** [review]

**Review by fkiraly — APPROVED** [review]
yes, all resolved from my side too. Nice contribution, thanks!

**Comment by fkiraly** [comment]
Did you add yourself to the all contributors file, @blazingbhavneek? (if not, I suggest separate PR)

**Pull request merged**